File: backend/core/responses/baseSongWithoutAlbumResponse.py
from typing import Sequence, Literal
from pydantic import BaseModel

from backend.core.responses.baseArtistResponse import BaseArtistResponse


class BaseSongWithoutAlbumResponse(BaseModel):
    """Base response model for an album's song information."""

    type: Literal["song"] = "song"
    provider: str
    publicId: str
    providerUrl: str
    name: str
    artists: Sequence[BaseArtistResponse]
    audioUrl: str | None
    downloaded: bool
    imageUrl: str
    dominantColor: str
    duration_ms: int
    discNumber: int
    trackNumber: int

File: frontend/dto/baseSongWithoutAlbumResponse.ts
import { BaseArtistResponseSchema } from "@/dto";
import { z } from "zod";

export const BaseSongWithoutAlbumResponseSchema = z.object({
    type: z.union([z.literal("song")]),
    provider: z.string(),
    publicId: z.string(),
    url: z.string(),
    name: z.string(),
    artists: z.array(z.lazy(() => BaseArtistResponseSchema)),
    audioSrc: z.string().nullable(),
    downloaded: z.boolean(),
    imageUrl: z.string(),
    duration: z.number(),
    discNumber: z.number(),
    trackNumber: z.number(),
});

export type BaseSongWithoutAlbumResponse = z.infer<
    typeof BaseSongWithoutAlbumResponseSchema
>;
